File: strands-ts/src/context-manager/retrieval-tool.ts
/**
 * Retrieval tool for accessing stashed (L1) content.
 *
 * Registered automatically when the ContextManager has storage configured.
 *
 * @internal
 */

import { z } from 'zod'
import { tool } from '../tools/tool-factory.js'
import { isSearchableContent, searchContent } from '../vended-plugins/context-offloader/search.js'
import type { JSONValue } from '../types/json.js'
import type { Tool } from '../tools/tool.js'
import { Message, ToolUseBlock } from '../types/messages.js'
import type { Stash } from './stash.js'

export const RETRIEVAL_TOOL_NAME = 'retrieve_context'

const DEFAULT_MAX_RESULT_TOKENS = 10_000
const CHARS_PER_TOKEN = 4

const retrievalInputSchema = z.object({
  reference: z.string().describe('The reference key from the offload placeholder.'),
  pattern: z.string().optional().describe('Regex or keyword to grep for. Returns matching lines with context.'),
  line_range: z
    .object({
      start: z.number().int().min(1).describe('First line to return (1-indexed).'),
      end: z.number().int().min(1).describe('Last line to return (1-indexed).'),
    })
    .optional()
    .describe('Return only this span of lines.'),
  context_lines: z.number().int().min(0).optional().describe('Lines of context around each match (default: 5).'),
})

/**
 * Creates the retrieval tool for the given stash.
 *
 * @param stash - The stash instance to retrieve from
 * @param maxResultTokens - Maximum tokens in the retrieval result
 * @returns A Tool that can retrieve stashed content
 *
 * @internal
 */
export function createRetrievalTool(stash: Stash, maxResultTokens?: number): Tool {
  const maxChars = (maxResultTokens ?? DEFAULT_MAX_RESULT_TOKENS) * CHARS_PER_TOKEN

  return tool({
    name: RETRIEVAL_TOOL_NAME,
    description:
      'Retrieve content that was offloaded from context.\n\n' +
      'When content is offloaded (truncated, dropped, or summarized), the original is ' +
      'persisted and a reference key is left in its place. Use this tool with that ' +
      'reference to access the original content.\n\n' +
      'Options:\n' +
      '  - pattern: regex/keyword to find matching lines with context\n' +
      '  - line_range: { start, end } to read a specific span\n' +
      '  - Without pattern/line_range: returns the full original content\n\n' +
      'Examples:\n' +
      '  { reference: "1_toolu_abc_0", pattern: "error" }\n' +
      '  { reference: "1_toolu_abc_0", line_range: { start: 10, end: 25 } }\n' +
      '  { reference: "1_toolu_abc_0" }',
    inputSchema: retrievalInputSchema,
    callback: async (input) => {
      const result = await stash.retrieve(input.reference)
      if (!result) {
        return `Error: reference not found: ${input.reference}`
      }

      if (!input.pattern && !input.line_range) {
        return result.data as JSONValue
      }

      const text = extractText(result.data)
      if (!text || !isSearchableContent('text/plain')) {
        return `Error: cannot search non-text content. Omit pattern/line_range to retrieve full content.`
      }

      const contextLines = input.context_lines ?? 5

      return searchContent(
        text,
        { pattern: input.pattern, line_range: input.line_range, context_lines: contextLines },
        maxChars
      )
    },
  })
}

/**
 * Tracks toolUseIds from retrieval tool invocations so their results
 * can be excluded from stashing (prevents retrieval loops).
 *
 * @internal
 */
export function trackRetrievalToolUseIds(message: Message, skipSet: Set<string>): void {
  if (message.role !== 'assistant') return
  for (const block of message.content) {
    if (block instanceof ToolUseBlock && block.name === RETRIEVAL_TOOL_NAME) {
      skipSet.add(block.toolUseId)
    }
  }
}

function extractText(data: unknown): string | null {
  if (typeof data === 'string') return data
  if (data && typeof data === 'object') {
    if ('text' in data && typeof (data as Record<string, unknown>).text === 'string') {
      return (data as Record<string, unknown>).text as string
    }
    if ('json' in data) {
      return JSON.stringify((data as Record<string, unknown>).json, null, 2)
    }
  }
  return null
}

File: strands-py/src/strands/_context_manager/retrieval_tool.py
"""Retrieval tool for accessing stashed (L1) content.

Registered automatically when the ContextManager has storage configured.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

from ..tools.tools import PythonAgentTool
from ..types.tools import ToolResult, ToolResultContent, ToolSpec
from ..vended_plugins.context_offloader.search import _is_searchable_content, _search_content

if TYPE_CHECKING:
    from .stash import Stash

logger = logging.getLogger(__name__)

RETRIEVAL_TOOL_NAME = "retrieve_context"

_DEFAULT_MAX_RESULT_TOKENS = 10_000
_CHARS_PER_TOKEN = 4


def _extract_text(data: object) -> str | None:
    """Extract searchable text from decoded stash data."""
    if isinstance(data, str):
        return data
    if isinstance(data, dict):
        text = data.get("text")
        if isinstance(text, str):
            return text
        json_val = data.get("json")
        if json_val is not None:
            return json.dumps(json_val, indent=2)
    return None


def _create_retrieval_tool(stash: Stash, max_result_tokens: int | None = None) -> PythonAgentTool:
    """Create the ``retrieve_context`` tool backed by the given stash."""
    max_chars = (max_result_tokens or _DEFAULT_MAX_RESULT_TOKENS) * _CHARS_PER_TOKEN

    async def _invoke(tool_use: dict[str, Any], **kwargs: Any) -> ToolResult:
        tool_use_id = tool_use["toolUseId"]
        inputs = tool_use.get("input", {})
        reference: str = inputs["reference"]
        pattern: str | None = inputs.get("pattern")
        line_range: dict[str, int] | None = inputs.get("line_range")
        context_lines: int | None = inputs.get("context_lines")

        result = await stash.retrieve(reference)
        if result is None:
            text = f"Error: reference not found: {reference}"
            return ToolResult(toolUseId=tool_use_id, status="error", content=[ToolResultContent(text=text)])

        if pattern is None and line_range is None:
            content: list[ToolResultContent] = [ToolResultContent(text=json.dumps(result))]
            return ToolResult(toolUseId=tool_use_id, status="success", content=content)

        text_content = _extract_text(result)
        if text_content is None or not _is_searchable_content("text/plain"):
            text = "Error: cannot search non-text content. Omit pattern/line_range to retrieve full content."
            return ToolResult(toolUseId=tool_use_id, status="error", content=[ToolResultContent(text=text)])

        ctx = context_lines if context_lines is not None else 5
        lr: tuple[int, int] | None = None
        if line_range is not None:
            lr = (int(line_range["start"]), int(line_range["end"]))

        search_result = _search_content(
            text_content, pattern=pattern, line_range=lr, context_lines=ctx, max_chars=max_chars
        )
        return ToolResult(toolUseId=tool_use_id, status="success", content=[ToolResultContent(text=search_result)])

    _invoke.__name__ = RETRIEVAL_TOOL_NAME

    tool_spec = ToolSpec(
        name=RETRIEVAL_TOOL_NAME,
        description=(
            "Retrieve offloaded content by reference.\n"
            "Usage modes:\n"
            "  - Full retrieval: { reference } — returns the complete stored content\n"
            "  - Pattern search: { reference, pattern, context_lines? } — grep for matches\n"
            "  - Line range: { reference, line_range: { start, end } } — extract lines (1-indexed)"
        ),
        inputSchema={
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "The reference key from the offload placeholder.",
                },
                "pattern": {
                    "type": "string",
                    "description": "Regex or keyword to grep for. Returns matching lines with context.",
                },
                "line_range": {
                    "type": "object",
                    "description": "Line range to extract (1-indexed, inclusive).",
                    "properties": {
                        "start": {"type": "integer", "minimum": 1, "description": "First line to return."},
                        "end": {"type": "integer", "minimum": 1, "description": "Last line to return."},
                    },
                    "required": ["start", "end"],
                },
                "context_lines": {
                    "type": "integer",
                    "minimum": 0,
                    "description": "Lines of context around each match (default: 5).",
                },
            },
            "required": ["reference"],
        },
    )

    return PythonAgentTool(
        tool_name=RETRIEVAL_TOOL_NAME,
        tool_spec=tool_spec,
        tool_func=_invoke,
    )


def _track_retrieval_tool_use_ids(message: dict[str, Any], skip_set: set[str]) -> None:
    """Track tool-use IDs from retrieve_context calls for loop prevention."""
    if message.get("role") != "assistant":
        return
    for block in message.get("content", []):
        if "toolUse" in block and block["toolUse"].get("name") == RETRIEVAL_TOOL_NAME:
            skip_set.add(block["toolUse"]["toolUseId"])
